clients/src/main/java/org/apache/kafka/common/protocol/RecordsWriter.java

```diff
@@ -0,0 +1,169 @@
+        this.dest = dest;
+        this.sendConsumer = sendConsumer;
+        this.byteArrayOutputStream = new ByteBufferOutputStream(32);
+        this.output = new DataOutputStream(this.byteArrayOutputStream);
+        this.mark = 0;
+    }
+
+    @Override
+    public void writeByte(byte val) {
+        try {
+            output.writeByte(val);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeShort(short val) {
+        try {
+            output.writeShort(val);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeInt(int val) {
+        try {
+            output.writeInt(val);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeLong(long val) {
+        try {
+            output.writeLong(val);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeDouble(double val) {
+        try {
+            ByteUtils.writeDouble(val, output);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeByteArray(byte[] arr) {
+        try {
+            output.write(arr);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeUnsignedVarint(int i) {
+        try {
+            ByteUtils.writeUnsignedVarint(i, output);
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeByteBuffer(ByteBuffer src) {
+        try {
+            output.write(src.array(), src.position(), src.remaining());
+        } catch (IOException e) {
+            throw new RuntimeException("RecordsWriter encountered an IO error", e);
+        }
+    }
+
+    public void writeRecords(BaseRecords records) {
+        flush();
+        sendConsumer.accept(records.toSend(dest));
+    }
+
+    /**
+     * Flush any pending bytes as a ByteBufferSend and reset the buffer
+     */
+    public void flush() {
+        ByteBuffer buf = byteArrayOutputStream.buffer();
+        int end = buf.position();
+        int len = end - mark;
+
+        if (len > 0) {
+            buf.position(mark);
+            ByteBuffer slice = buf.slice();
+            slice.limit(len);
+            ByteBufferSend send = new ByteBufferSend(dest, slice);
+            sendConsumer.accept(send);
+        }
+
+        buf.position(end);
+        mark = end;
```

Comment:
I ended up having to keep a separate mark here since ByteBufferOutputStream doesn't keep the mark when it replaces the underlying buffer.

Reply:
Ended up not using ByteBufferOutputStream